File: UserService/src/main/scala/APIs/UserService/SendMessageMessage.scala
package APIs.UserService

import Common.API.API
import Global.ServiceCenter.UserServiceCode

import io.circe.{Decoder, Encoder, Json}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import io.circe.syntax.*
import io.circe.parser.*
import Common.Serialize.CustomColumnTypes.{decodeDateTime,encodeDateTime}

import com.fasterxml.jackson.core.`type`.TypeReference
import Common.Serialize.JacksonSerializeUtils

import scala.util.Try

import org.joda.time.DateTime
import java.util.UUID

/**
 * SendMessageMessage
 * desc: 向指定用户发送消息，将消息存储到双方的message_box中。
 * @param userToken: String (发送者的用户凭证，用于验证用户身份。)
 * @param recipientID: String (接收者的用户ID。)
 * @param messageContent: String (消息内容。)
 * @return result: String (发送结果，例如'消息发送成功！')
 */

case class SendMessageMessage(
  userToken: String,
  recipientID: String,
  messageContent: String
) extends API[String](UserServiceCode)



case object SendMessageMessage{
    
  import Common.Serialize.CustomColumnTypes.{decodeDateTime,encodeDateTime}

  // Circe 默认的 Encoder 和 Decoder
  private val circeEncoder: Encoder[SendMessageMessage] = deriveEncoder
  private val circeDecoder: Decoder[SendMessageMessage] = deriveDecoder

  // Jackson 对应的 Encoder 和 Decoder
  private val jacksonEncoder: Encoder[SendMessageMessage] = Encoder.instance { currentObj =>
    Json.fromString(JacksonSerializeUtils.serialize(currentObj))
  }

  private val jacksonDecoder: Decoder[SendMessageMessage] = Decoder.instance { cursor =>
    try { Right(JacksonSerializeUtils.deserialize(cursor.value.noSpaces, new TypeReference[SendMessageMessage]() {})) } 
    catch { case e: Throwable => Left(io.circe.DecodingFailure(e.getMessage, cursor.history)) }
  }
  
  // Circe + Jackson 兜底的 Encoder
  given sendMessageMessageEncoder: Encoder[SendMessageMessage] = Encoder.instance { config =>
    Try(circeEncoder(config)).getOrElse(jacksonEncoder(config))
  }

  // Circe + Jackson 兜底的 Decoder
  given sendMessageMessageDecoder: Decoder[SendMessageMessage] = Decoder.instance { cursor =>
    circeDecoder.tryDecode(cursor).orElse(jacksonDecoder.tryDecode(cursor))
  }
}

File: frontend/src/Plugins/UserService/APIs/SendMessageMessage.ts
/**
 * SendMessageMessage
 * desc: 向指定用户发送消息，将消息存储到双方的message_box中。
 * @param userToken: String (发送者的用户凭证，用于验证用户身份。)
 * @param recipientID: String (接收者的用户ID。)
 * @param messageContent: String (消息内容。)
 * @return result: String (发送结果，例如'消息发送成功！')
 */
import { TongWenMessage } from '../../TongWenAPI/TongWenMessage'
import { ServiceConfig } from '../../../Globals/ServiceConfig'



export class SendMessageMessage extends TongWenMessage {
    constructor(
        public  userToken: string,
        public  recipientID: string,
        public  messageContent: string
    ) {
        super()
    }
    getAddress(): string {
        return ServiceConfig.getUserServiceAddress()
    }
}

